client/client.go

```diff
@@ -99,56 +123,513 @@ func UploadAdd(issuer principal.Signer, space did.DID, params uploadadd.Caveat,
+			}
+
+			acceptOk, err := result.Unwrap(result.MapError(acceptRcpt.Out(), failure.FromFailureModel))
+			if err != nil {
+				return nil, nil, fmt.Errorf("blob/accept failed: %w", err)
+			}
+
+			site = acceptOk.Site
+			rcptBlocks = acceptRcpt.Blocks()
+		} else {
+			legacyAcceptRcpt, err = receipt.Rebind[w3sBlob.AcceptOk, fdm.FailureModel](anyAcceptRcpt, w3sBlob.AcceptOkType(), fdm.FailureType(), captypes.Converters...)
+			if err != nil {
+				return nil, nil, fmt.Errorf("fetching legacy accept receipt: %w", err)
+			}
+
+			acceptOk, err := result.Unwrap(result.MapError(legacyAcceptRcpt.Out(), failure.FromFailureModel))
+			if err != nil {
+				return nil, nil, fmt.Errorf("web3.storage/blob/accept failed: %w", err)
+			}
+
+			site = acceptOk.Site
+			rcptBlocks = legacyAcceptRcpt.Blocks()
+		}
+	}
+
+	locationBlocks, err := blockstore.NewBlockStore(blockstore.WithBlocksIterator(rcptBlocks))
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading location commitment blocks: %w", err)
+	}
+
+	location, err := delegation.NewDelegationView(site, locationBlocks)
+	if err != nil {
+		return nil, nil, fmt.Errorf("creating location delegation: %w", err)
+	}
+
+	return contentLink, location, nil
+}
+
+func getConcludeReceipt(concludeFx invocation.Invocation) (receipt.AnyReceipt, error) {
+	concludeNb, fail := ucancap.ConcludeCaveatsReader.Read(concludeFx.Capabilities()[0].Nb())
+	if fail != nil {
+		return nil, fmt.Errorf("invalid conclude receipt: %w", fail)
+	}
+
+	reader := receipt.NewAnyReceiptReader(captypes.Converters...)
+	rcpt, err := reader.Read(concludeNb.Receipt, concludeFx.Blocks())
+	if err != nil {
+		return nil, fmt.Errorf("reading receipt: %w", err)
+	}
+
+	return rcpt, nil
+}
+
+func putBlob(url url.URL, headers http.Header, body []byte) error {
+	req, err := http.NewRequest(http.MethodPut, url.String(), bytes.NewReader(body))
+	if err != nil {
+		return fmt.Errorf("creating upload request: %w", err)
+	}
+
+	for k, v := range headers {
+		req.Header.Set(k, v[0])
+	}
+
+	// TODO: custom HTTP client with timeout
+	// TODO: retries
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("uploading blob: %w", err)
+	}
+	io.ReadAll(resp.Body)
+	resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("uploading blob: %s", resp.Status)
+	}
+
+	return nil
+}
+
+func sendPutReceipt(putTask invocation.Invocation, issuer ucan.Signer, audience ucan.Principal, conn client.Connection) error {
+	if len(putTask.Facts()) != 1 {
+		return fmt.Errorf("invalid put facts, wanted 1 fact but got %d", len(putTask.Facts()))
+	}
+
+	if _, ok := putTask.Facts()[0]["keys"]; !ok {
+		return fmt.Errorf("invalid put facts, missing 'keys' field")
+	}
+
+	putKeysNode, ok := putTask.Facts()[0]["keys"].(ipld.Node)
+	if !ok {
+		return fmt.Errorf("invalid put facts, 'keys' field is not a node")
+	}
+
+	// TODO: define a schema and use bindnode.Rebind rather than doing this manually
```

Comment:
yea that will make you a happier camper :P dealing with raw ipld.Node interfaces is a mess